src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+
+        etot_old = energy_total
+        energy_total = info.energies.total
+        abs(energy_total - etot_old) < tolerance
+    end
+    return is_converged
+end
+:w
+
+"""
+Flag convergence as soon as density change drops below tolerance
+"""
+function NewtonConvergenceDensity(tolerance)
+    info -> (norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol) < tolerance)
+end
+
+"""
+Default callback function for `newton`, which prints a convergence table
+"""
+function NewtonDefaultCallback()
+    prev_energies = nothing
+    function callback(info)
+        !mpi_master() && return info  # Printing only on master
+        if info.stage == :finalize
+            info.converged || @warn "Newton not converged."
+            return info
+        end
+        collinear = info.basis.model.spin_polarization == :collinear
+
+        if info.n_iter == 1
+            E_label = haskey(info.energies, "Entropy") ? "Free energy" : "Energy"
+            magn    = collinear ? ("   Magnet", "   ------") : ("", "")
+            @printf "n     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ%s\n" E_label magn[1]
+            @printf "---   ---------------   ---------   --------%s\n" magn[2]
+        end
+        E    = isnothing(info.energies) ? Inf : info.energies.total
+        Δρ   = norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol)
+        if size(info.ρ_next, 4) == 1
+            magn = NaN
+        else
+            magn = sum(spin_density(info.ρ_next)) * info.basis.dvol
+        end
+
+        Estr   = (@sprintf "%+15.12f" round(E, sigdigits=13))[1:15]
+        prev_E = prev_energies === nothing ? Inf : prev_energies.total
+        ΔE     = prev_E == Inf ? "      NaN" : @sprintf "% 3.2e" E - prev_E
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+#
+# Newton algorithm
+#
+
+"""
+    newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                tol_krylov=1e-12, krylov_verbosity=1)
+
+Perform a newton step : we take as given a planewave set φ and we return the
+newton step δφ and the updated state φ - δφ
+(after proper orthonormalization) where δφ solves Jac * δφ = res
+and Jac is the Jacobian of the projected gradient descent : Jac = Ω+K.
+δφ is an element of the tangent space at φproj (set to φ if not specified in newton function).
+"""
+function newton_step(basis::PlaneWaveBasis, φ, φproj, res, occ;
+                     tol_krylov=1e-12, krylov_verbosity=1)
+
+    # necessary quantities
+    N = size(φ[1], 2)
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρproj = compute_density(basis, φproj, occ)
+    energies, Hproj = energy_hamiltonian(basis, φproj, occ; ρ=ρproj)
+
+    # packing routines
+    pack(φ) = pack_arrays(basis, φ)
```

Comment:
I'd just call them directly

Reply:
Same remark than for direct minimization.